File: demo/repo_map.py
# demo/repo_map.py
"""The repository map: what Icarus has INDEXED, said before anyone asks.

This is the first surface that speaks first, which makes it the first surface
that can over-claim. The whole design constraint follows from one honest
distinction:

    a corpus-derived map describes what Icarus READ.
    It does not describe what EXISTS in the repository.

Nothing here is a claim about the customer's repo. Every field is named
`indexed_*`, every count is derived from refs the corpus actually holds, and the
ingest deny-lists are reported as RULES THAT WERE APPLIED rather than as
evidence about particular files -- because `evals.ingest.classify_file` returns
None and the walk moves on, recording nothing about what it skipped. Publishing
an excluded-file count would be fabricated precision, which is the same class of
failure as a bluffed citation.

Pure: chunks in, dict out. No model call, no network, no filesystem -- the
corpus is already in memory (`GatedPipeline.indexed_chunks`), so a 50k-chunk
repo costs no re-read. A future ingestion-manifest brick can add genuinely OBSERVED discovered
/eligible/excluded/failed counts; until then those numbers do not exist and are
not published.
"""

from evals.ingest import (
    _DENY_DIR_SEGMENTS,
    _DENY_FILENAMES,
    _DENY_FILENAME_SUFFIXES,
    _MAX_FILE_BYTES,
)

from .entry_points import detect_entry_points, is_auxiliary_path

# Sources whose ref carries a repo-relative PATH (`code:llm/cli.py#L1-L300`).
# pr/issue/commit refs carry an identifier instead and have no file at all.
_FILE_SOURCES = ("code", "doc", "config")

# Extension -> the name a human would use. Display only: it never affects
# ingest, retrieval or the gate. An unmapped extension falls back to the
# extension itself, so a file can never vanish from the totals just because
# this table doesn't know its language (the totals must equal
# indexed_file_count, and a test pins that).
_LANGUAGE_BY_EXT = {
    ".py": "Python", ".js": "JavaScript", ".jsx": "JavaScript",
    ".mjs": "JavaScript", ".cjs": "JavaScript", ".ts": "TypeScript",
    ".tsx": "TypeScript", ".go": "Go", ".rs": "Rust", ".java": "Java",
    ".rb": "Ruby", ".c": "C", ".h": "C/C++ header", ".cpp": "C++",
    ".m": "Objective-C", ".mm": "Objective-C++", ".swift": "Swift",
    ".kt": "Kotlin", ".php": "PHP", ".cs": "C#", ".scala": "Scala",
    ".sh": "Shell", ".md": "Markdown", ".rst": "reStructuredText",
    ".txt": "Text", ".yaml": "YAML", ".yml": "YAML", ".toml": "TOML",
    ".cfg": "Config", ".ini": "Config", ".sql": "SQL",
    ".gradle": "Gradle", ".podspec": "Podspec",
}

_ROOT = "."  # the directory bucket for a file with no directory component


def _split(ref):
    """`code:llm/cli.py#L1-L300` -> ("code", "llm/cli.py"). A ref with no path
    component (pr/issue/commit) returns its source and None."""
    source, _, rest = ref.partition(":")
    if source not in _FILE_SOURCES or not rest:
        return source, None
    return source, rest.split("#", 1)[0]


def _language(path):
    dot = path.rfind(".")
    ext = path[dot:].lower() if dot > path.rfind("/") else ""
    return _LANGUAGE_BY_EXT.get(ext, ext or "(no extension)")


def _top_directory(path):
    head, sep, _ = path.partition("/")
    return head if sep else _ROOT


def _readme(doc_paths):
    """The shallowest README, or None. Sorted by (depth, path) so the choice is
    deterministic and a root README beats a nested one. Never falls back to
    "some other doc that looks a bit like a readme" -- a missing README is a
    fact about the corpus and is reported as one."""
    candidates = [p for p in doc_paths if p.rsplit("/", 1)[-1].lower().startswith("readme")]
    if not candidates:
        return None
    return sorted(candidates, key=lambda p: (p.count("/"), p))[0]


def _exclusion_rules():
    """The rules ingest applied, derived from ingest's own constants so they
    cannot drift out of sync with what the walk really does. These describe
    POLICY, not observations."""
    return [
        f"Directories skipped anywhere in a path: {', '.join(sorted(_DENY_DIR_SEGMENTS))}.",
        f"Filenames skipped: {', '.join(sorted(_DENY_FILENAMES))}.",
        f"Filename suffixes skipped: {', '.join(sorted(_DENY_FILENAME_SUFFIXES))}.",
        "Only an allow-listed set of source, documentation and configuration "
        "extensions is indexed; other extensions (including .json) are not.",
        f"Any single file larger than {_MAX_FILE_BYTES // 1024} KB is skipped.",
        "Files whose first bytes contain a null byte are treated as binary and skipped.",
    ]


def build_map(chunks, status):
    """The repository map for one connected repo.

    `chunks` is the loaded corpus (`GatedPipeline.indexed_chunks()`); `status`
    is a `Library.status_snapshot()`. Both are already in memory, so this is
    cheap enough to serve per request -- nothing is re-read from disk.

    Chunks rather than refs because entry-point detection has to READ a
    manifest (`pyproject.toml`'s declared console scripts) and a file's own
    text; refs alone cannot carry that. Everything else here still needs only
    the ref.
    """
    paths, chunks_by_source = {}, {}
    for chunk in chunks:
        source, path = _split(chunk.ref)
        chunks_by_source[source] = chunks_by_source.get(source, 0) + 1
        if path:
            paths[path] = source  # distinct FILE, however many chunks it made

    languages, directories = {}, {}
    for path in paths:
        lang = _language(path)
        languages[lang] = languages.get(lang, 0) + 1
        top = _top_directory(path)
        directories[top] = directories.get(top, 0) + 1

    docs = sorted(p for p, s in paths.items() if s == "doc")
    truncated = bool(status.get("truncated", False))
    entry_points = detect_entry_points(chunks)
    # Committed test data is LABELLED beside the totals, never subtracted from
    # them. Found by running the tour on this repo: 348 of 500 indexed files
    # sat under `evals/fixtures/`, so the map read as an Objective-C++/Kotlin/
    # Swift mobile codebase. Every number was true and the impression was
    # false -- the same failure this product exists to avoid, moved out of
    # answers and into a summary. Excluding them would "fix" it by breaking
    # the map's one contract: that it describes what Icarus READ.
    auxiliary = sum(1 for p in paths if is_auxiliary_path(p))

    limitations = [
        "This map describes what Icarus INDEXED, not what exists in the "
        "repository. It cannot report how many files the repository has.",
        "The exclusion rules below are the rules that were applied. Icarus did "
        "not record which particular files they skipped, so it cannot say how "
        "many files were excluded or name them.",
        "Entry points are found by explicit rules only, and each one names the "
        "rule and the evidence that produced it. A repository whose entry point "
        "no rule recognises will show none rather than a guess -- and "
        "package.json is not indexed, so JavaScript and TypeScript entry points "
        "are recognised by conventional filename only.",
    ]
    if truncated:
        limitations.append(
            "A size cap truncated this ingest, so parts of the repository were "
            "not read at all. Anything Icarus says about coverage stops here.")
    if status.get("indexing"):
        limitations.append(
            "Semantic indexing is still running. Search works, but it is "
            "keyword-only for now and answers are measurably worse until it "
            "finishes -- an abstention right now may mean Icarus has not "
            "finished reading, not that nobody wrote it down.")

    return {
        "repo": status.get("repo"),
        "commit": status.get("commit"),
        "indexed_file_count": len(paths),
        "indexed_languages": dict(sorted(languages.items(), key=lambda kv: (-kv[1], kv[0]))),
        "indexed_directories": dict(sorted(directories.items(), key=lambda kv: (-kv[1], kv[0]))),
        "indexed_documentation": {"files": docs, "readme": _readme(docs)},
        "indexed_entry_points": entry_points,
        "indexed_auxiliary": {
            "file_count": auxiliary,
            "rule": "Counted within the totals above, and named here because a "
                    "large committed test tree makes a repository look like "
                    "something it is not: files under a test, tests, testing, "
                    "spec, fixture, fixtures, testdata or __fixtures__ "
                    "directory, or named test_*/*_test. The same rule keeps "
                    "them out of the entry points.",
        },
        "indexed_chunks_by_source": dict(sorted(chunks_by_source.items())),
        "lexical_search_ready": bool(chunks_by_source),
        "semantic_indexing_in_progress": bool(status.get("indexing", False)),
        "corpus_truncated": truncated,
        "exclusion_rules": _exclusion_rules(),
        "limitations": limitations,
    }

File: mac/Icarus/Sources/IcarusKit/RepoMap.swift
import Foundation

/// `GET /map` — what Icarus has INDEXED for the connected repo.
///
/// Every field is named `indexed*` for the same reason the brain names them
/// that way (`demo/repo_map.py`): a map built from the corpus describes what
/// Icarus READ, never what EXISTS in the repository. There is deliberately no
/// total-file count and no excluded-file count here, because the brain does not
/// know them and inventing them would be the same failure as a bluffed
/// citation. `exclusionRules` are the rules that were APPLIED, not observations
/// about particular files.
public struct RepoMap: Decodable, Sendable {
    public let repo: String?
    public let commit: String?
    public let indexedFileCount: Int
    /// Distinct FILES per language (never chunks — a file can produce many).
    public let indexedLanguages: [String: Int]
    /// Distinct files per top-level directory ("." for repo-root files).
    public let indexedDirectories: [String: Int]
    public let indexedDocumentation: IndexedDocumentation
    public let indexedEntryPoints: [EntryPoint]
    /// Committed test data, counted WITHIN the totals above and named
    /// separately. Optional so a brain deployed before this field existed
    /// still decodes — absent simply means the share is unknown, never zero.
    public let indexedAuxiliary: IndexedAuxiliary?
    public let indexedChunksBySource: [String: Int]
    public let lexicalSearchReady: Bool
    public let semanticIndexingInProgress: Bool
    public let corpusTruncated: Bool
    public let exclusionRules: [String]
    public let limitations: [String]
    /// How the code is ARRANGED, derived from its own imports
    /// (`demo/structure.py`). Optional so a brain deployed before this field
    /// existed still decodes -- absent means Icarus did not report an
    /// arrangement, never that the repository has none.
    public let indexedStructure: IndexedStructure?

    enum CodingKeys: String, CodingKey {
        case repo, commit, limitations
        case indexedStructure = "indexed_structure"
        case indexedFileCount = "indexed_file_count"
        case indexedLanguages = "indexed_languages"
        case indexedDirectories = "indexed_directories"
        case indexedDocumentation = "indexed_documentation"
        case indexedEntryPoints = "indexed_entry_points"
        case indexedAuxiliary = "indexed_auxiliary"
        case indexedChunksBySource = "indexed_chunks_by_source"
        case lexicalSearchReady = "lexical_search_ready"
        case semanticIndexingInProgress = "semantic_indexing_in_progress"
        case corpusTruncated = "corpus_truncated"
        case exclusionRules = "exclusion_rules"
    }

    public init(repo: String?, commit: String?, indexedFileCount: Int,
                indexedLanguages: [String: Int], indexedDirectories: [String: Int],
                indexedDocumentation: IndexedDocumentation,
                indexedEntryPoints: [EntryPoint], indexedAuxiliary: IndexedAuxiliary? = nil,
                indexedChunksBySource: [String: Int],
                lexicalSearchReady: Bool, semanticIndexingInProgress: Bool,
                corpusTruncated: Bool, exclusionRules: [String], limitations: [String],
                indexedStructure: IndexedStructure? = nil) {
        self.indexedStructure = indexedStructure
        self.repo = repo
        self.commit = commit
        self.indexedFileCount = indexedFileCount
        self.indexedLanguages = indexedLanguages
        self.indexedDirectories = indexedDirectories
        self.indexedDocumentation = indexedDocumentation
        self.indexedEntryPoints = indexedEntryPoints
        self.indexedAuxiliary = indexedAuxiliary
        self.indexedChunksBySource = indexedChunksBySource
        self.lexicalSearchReady = lexicalSearchReady
        self.semanticIndexingInProgress = semanticIndexingInProgress
        self.corpusTruncated = corpusTruncated
        self.exclusionRules = exclusionRules
        self.limitations = limitations
    }
}

/// How much of what Icarus read is supporting material rather than the project.
///
/// A large committed fixture tree makes a repository look like something it is
/// not — measured on this repo, 348 of 500 indexed files sat under
/// `evals/fixtures/`, so the language mix read as a mobile codebase. The count
/// sits BESIDE the totals rather than being subtracted from them, because the
/// map's one contract is that it describes what Icarus READ.
public struct IndexedAuxiliary: Decodable, Sendable {
    public let fileCount: Int
    /// The rule that produced the count, in words — never an opaque number.
    public let rule: String

    enum CodingKeys: String, CodingKey {
        case rule
        case fileCount = "file_count"
    }

    public init(fileCount: Int, rule: String) {
        self.fileCount = fileCount
        self.rule = rule
    }
}

public struct IndexedDocumentation: Decodable, Sendable {
    public let files: [String]
    /// nil when NO readme was indexed. Rendered as an explicit "none was
    /// indexed", never as silence — the absence is a fact about the corpus.
    public let readme: String?

    public init(files: [String], readme: String?) {
        self.files = files
        self.readme = readme
    }
}

/// One file the brain believes is a place to start reading, with the RULE that
/// produced it. Nothing here is scored or ranked: a rule fired, or it didn't.
public struct EntryPoint: Decodable, Identifiable, Sendable {
    public let path: String
    public let rules: [EntryPointRule]
    public var id: String { path }

    public init(path: String, rules: [EntryPointRule]) {
        self.path = path
        self.rules = rules
    }
}

public struct EntryPointRule: Decodable, Identifiable, Sendable {
    public let rule: String
    /// The indexed chunk that proves it — citable like any other evidence.
    public let evidenceRef: String?
    public let detail: String

    public var id: String { rule }

    enum CodingKeys: String, CodingKey {
        case rule, detail
        case evidenceRef = "evidence_ref"
    }

    public init(rule: String, evidenceRef: String?, detail: String) {
        self.rule = rule
        self.evidenceRef = evidenceRef
        self.detail = detail
    }
}

public extension RepoMap {
    /// Languages biggest-first, for display. Ties break on name so the order is
    /// stable between renders rather than dictionary-order roulette.
    var languagesByFileCount: [(name: String, files: Int)] {
        indexedLanguages.sorted { ($0.value, $1.key) > ($1.value, $0.key) }
            .map { (name: $0.key, files: $0.value) }
    }

    var directoriesByFileCount: [(name: String, files: Int)] {
        indexedDirectories.sorted { ($0.value, $1.key) > ($1.value, $0.key) }
            .map { (name: $0.key, files: $0.value) }
    }
}

/// How the code is ARRANGED, read off its own import statements
/// (`demo/structure.py`).
///
/// This answers the question measurement found hardest — the writer-backed
/// `architecture` step scored 2 of 10 real repositories, and anchoring it to a
/// README provably could not fix it, because a README says what a project is
/// FOR, not how its code is laid out. So it is derived rather than asked, and
/// it cannot bluff: every edge carries the indexed chunk that proves it.
///
/// **An empty arrangement is not a claim that there is none.** Only Python,
/// JavaScript/TypeScript and Go imports are analysed; anything else is listed
/// in `unanalysedLanguages`, and a view MUST say so rather than showing an
/// empty panel that reads as "this project has no structure".
public struct IndexedStructure: Decodable, Equatable, Sendable {
    public let components: [Component]
    public let mostDependedOnFiles: [CoreFile]
    /// Imports seen and not resolved to another indexed file. Mostly
    /// third-party and standard-library imports, which are external by
    /// definition — a large number here is normal and does not mean anything
    /// is missing.
    public let unresolvedImportCount: Int
    /// Indexed languages with no resolver at all. Naming them is what stops an
    /// empty result reading as "no structure exists".
    public let unanalysedLanguages: [String]
    public let limitations: [String]

    private enum CodingKeys: String, CodingKey {
        case components, limitations
        case mostDependedOnFiles = "most_depended_on_files"
        case unresolvedImportCount = "unresolved_import_count"
        case unanalysedLanguages = "unanalysed_languages"
    }

    public struct Component: Decodable, Equatable, Sendable {
        public let path: String
        public let fileCount: Int
        public let dependsOn: [String]
        public let dependedOnBy: [String]
        /// The indexed chunks proving this component's edges — citable like
        /// any other claim Icarus makes.
        public let evidenceRefs: [String]

        private enum CodingKeys: String, CodingKey {
            case path
            case fileCount = "file_count"
            case dependsOn = "depends_on"
            case dependedOnBy = "depended_on_by"
            case evidenceRefs = "evidence_refs"
        }

        public init(path: String, fileCount: Int, dependsOn: [String],
                    dependedOnBy: [String], evidenceRefs: [String]) {
            self.path = path
            self.fileCount = fileCount
            self.dependsOn = dependsOn
            self.dependedOnBy = dependedOnBy
            self.evidenceRefs = evidenceRefs
        }
    }

    public struct CoreFile: Decodable, Equatable, Sendable {
        public let path: String
        public let dependedOnByCount: Int
        public let evidenceRef: String

        private enum CodingKeys: String, CodingKey {
            case path
            case dependedOnByCount = "depended_on_by_count"
            case evidenceRef = "evidence_ref"
        }

        public init(path: String, dependedOnByCount: Int, evidenceRef: String) {
            self.path = path
            self.dependedOnByCount = dependedOnByCount
            self.evidenceRef = evidenceRef
        }
    }

    public init(components: [Component], mostDependedOnFiles: [CoreFile],
                unresolvedImportCount: Int, unanalysedLanguages: [String],
                limitations: [String]) {
        self.components = components
        self.mostDependedOnFiles = mostDependedOnFiles
        self.unresolvedImportCount = unresolvedImportCount
        self.unanalysedLanguages = unanalysedLanguages
        self.limitations = limitations
    }

    /// Did Icarus find an arrangement at all?
    public var isEmpty: Bool { components.isEmpty && mostDependedOnFiles.isEmpty }

    /// What to show when nothing was found — which is never "no structure".
    public var emptyExplanation: String {
        if !unanalysedLanguages.isEmpty {
            return "Icarus doesn’t analyse \(unanalysedLanguages.joined(separator: ", ")) "
                 + "imports, so it didn’t look for structure here. That isn’t a claim "
                 + "that there is none."
        }
        return "No internal imports resolved, so there is nothing to show. That can be "
             + "true of a project whose code lives in one flat package."
    }
}
